File: mpi-1x-hypack-2013/collective-comm-comp-codes/f77-lang-collect-mpi/allreduce.f
c
c******************************************************************
c
c		C-DAC Tech Workshop : hyPACK-2013
c                           October 15-18, 2013
c
c    Example 2.7	 : Allreduce.f 
c
c    Objective           : To broadcast an integer array of size "n" by process with rank "0"  
c                          and perform Global summation using MPI_Allreduce
c                          (Combine values from all processes and distributed the result
c                           back to all processes)
c                        
c                          This example demonstrates the use of
c                          MPI_Init
c                          MPI_Comm_rank
c                          MPI_Comm_size
c                          MPI_Bast
c                          MPI_Allreduce                      
c                          MPI_Finalize
c
c    Input               : Input Data file "sdata.inp" by proces with Rank 0
c
c    Output              : Print the final sum on all processes.
c
c    Necessary Condition : Number of processes should be 
c                          less than or equal to 8.
c
c   Created             : August-2013
c
c   E-mail              : hpcfte@cdac.in     
c
c*******************************************************************

	   program main

	   include "mpif.h"

         integer SendSumValue, RecvSumValue
	 integer Numprocs, MyRank, Root
         integer count, ierror

C .........MPI Initialisation .......
	 call MPI_INIT( ierror )
  	 call MPI_COMM_RANK(MPI_COMM_WORLD, MyRank,  ierror)
  	 call MPI_COMM_SIZE(MPI_COMM_WORLD, Numprocs,ierror)

  	 Root = 0
         if( MyRank .eq. Root) then
    	     open (unit= 18, FILE = "./data/SendSumValue.inp")
    	       read(18,*) SendSumValue
    	        print*, MyRank, SendSumValue
	   endif
c 
        call MPI_Bcast(SendSumValue,1,MPI_INTEGER,Root,
     $                  MPI_COMM_WORLD,ierror)

c .....Allreduce operation ....
         
        count = 1
        call MPI_Allreduce(SendSumValue, RecvSumValue, count, 
     $ MPI_INTEGER, MPI_SUM, MPI_COMM_WORLD, ierror) 

       print*,"MyRank = ", MyRank, "Sum =", RecvSumValue
 	 
      call MPI_FINALIZE(ierror)

	stop
	end

File: mpi-1x-hypack-2013/collective-comm-comp-codes/c-lang-collect-mpi/allreduce.c

/******************************************************************

		C-DAC Tech Workshop : hyPACK-2013
                           October 15-18, 2013

    Example 2.7		: Allreduce.c

    Objective           : To broadcast an integer array of size "n" by 
                          process with rank "0"  and perform Global 
                          summation using MPI_Allreduce (Combine values 
                          from all processes and distributed the result
                          back to all processes)
                        
                          This example demonstrates the use of

                          MPI_Init
                          MPI_Comm_rank
                          MPI_Comm_size
                          MPI_Bast
                          MPI_Allreduce                      
                          MPI_Finalize

    Input               : Input Data file "sdata.inp" by proces with Rank 0

    Output              : Print the final sum on all processes.

    Necessary Condition : Number of processes should be 
                          less than or equal to 8.

   Created             : August-2013

   E-mail              : hpcfte@cdac.in     

*******************************************************************/

#include<stdio.h>
#include "mpi.h"
main(int argc ,char **argv)
{
	int rank, size;
 	int no,sum;	


	MPI_Status status;

	MPI_Init(&argc,&argv);
	MPI_Comm_rank(MPI_COMM_WORLD,&rank);
	MPI_Comm_size(MPI_COMM_WORLD,&size);
	
	no=rank;		
	MPI_Allreduce(&no,&sum,2,MPI_INT,MPI_SUM,MPI_COMM_WORLD);

	
	printf("Final sum is=  %d\n",sum);
	
	MPI_Finalize();
}

			



		
	

	








		

